Lazy save 2

pp-mo opened the pull request:
## 🚀 Pull Request

### Description
<!-- Provide a clear description about your awesome pull request -->
<!-- Tell us all about your new feature, improvement, or bug fix -->


---
[Consult Iris pull request check list]( https://scitools-iris.readthedocs.io/en/latest/developers_guide/contributing_pull_request_checklist.html)

pp-mo commented:
Although there are no tests yet, I have created a simple routine to demonstrate function.
In case that may be useful, I've put it here : https://gist.github.com/pp-mo/a5b1d693c4a27192e5e0e2c2c4fa5ca9

Possibly a bit poor, but there it is !

pp-mo commented:
:eyes: @schlunma @zklaus @bouweandela @fnattino @trexfeathers

pp-mo commented:
### Problem with fill-value checking

I hit a glitch here due to the different behaviour of dask execution with a distributed scheduler.
With a _distributed_ scheduler, it is no longer possible to do direct streaming of variable data (i.e. `da.store(compute=True)`) ***while the caller holds the file open for writing***, as is wanted [here](https://github.com/SciTools/iris/blob/cf611f7d07b9980171ce681ce51f359876f05b8e/lib/iris/fileformats/netcdf/saver.py#L2524-L2532).
That's because, it's asking another process to open the file for write, while we already have it open for write by the caller.  
( And, we probably don't want to close/reopen it since that will invalidate any existing netCDF4 data objects which the Saver code probably stores + expects to continue working with ?   I'm not actually sure of this, though.  )

So for now, I've completely fudged the issue:
  1. simply ignore the fill-value check [for lazy operations](https://github.com/SciTools/iris/blob/cf611f7d07b9980171ce681ce51f359876f05b8e/lib/iris/fileformats/netcdf/saver.py#L2519-L2521)
  2. for-now actioning [**_all_** saves as lazy](https://github.com/SciTools/iris/blob/cf611f7d07b9980171ce681ce51f359876f05b8e/lib/iris/fileformats/netcdf/saver.py#L2856), and simply ['completing' the non-lazy ones](https://github.com/SciTools/iris/blob/cf611f7d07b9980171ce681ce51f359876f05b8e/lib/iris/fileformats/netcdf/saver.py#L2856).

This presents us with a bit of a backwards-compatibilty problem, in that I don't see how we can't check the data for fill-value collisions in the "new world".  And while that may clearly be okay when using the new lazy-saving, I'm less comfortable that it just won't work for

pp-mo commented:
Changed message on previous commit, and added a fix.

There is one remaining test failure, but I'm suspicious that this might be a sporadic problem :scream: 
The log looks like this ...
```
2023-03-10T16:58:50.6415706Z _________________ TestNetCDFSave.test_netcdf_save_multi2single _________________
2023-03-10T16:58:50.6416608Z [gw1] linux -- Python 3.9.16 /home/runner/work/iris/iris/.nox/tests/bin/python
2023-03-10T16:58:50.6416835Z 
2023-03-10T16:58:50.6417058Z self = <iris.tests.test_netcdf.TestNetCDFSave testMethod=test_netcdf_save_multi2single>
2023-03-10T16:58:50.6417310Z 
2023-03-10T16:58:50.6417410Z     @tests.skip_data
2023-03-10T16:58:50.6417678Z     def test_netcdf_save_multi2single(self):
2023-03-10T16:58:50.6418118Z         # Test saving multiple cubes to a single CF-netCDF file.
2023-03-10T16:58:50.6418420Z         # Read PP input file.
2023-03-10T16:58:50.6418689Z         file_in = tests.get_data_path(
2023-03-10T16:58:50.6419014Z             ("PP", "cf_processing", "abcza_pa19591997_daily_29.b.pp")
2023-03-10T16:58:50.6419284Z         )
2023-03-10T16:58:50.6419789Z         cubes = iris.load(file_in)
2023-03-10T16:58:50.6420020Z     
2023-03-10T16:58:50.6420253Z         # Write Cube to netCDF file.
2023-03-10T16:58:50.6420570Z         with self.temp_filename(suffix=".nc") as file_out:
2023-03-10T16:58:50.6420888Z             # Check that it is the same on loading
2023-03-10T16:58:50.6421163Z >           iris.save(cubes, file_out)
2023-03-10T16:58:50.6421342Z 
2023-03-10T16:58:50.6421465Z lib/iris/tests/test_netcdf.py:728: 
2023-03-10T16:58:50.6421815Z _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
2023-03-10T16:58:50.6422089Z lib/iris/io/__init__.py:485: in save
2023-03-10T16:58:50.6422390Z     result = saver(source, target, **kwargs)
2023-03-10T16:58:50.6422804Z lib/iris/fileformats/netcdf/saver.py:2906: in save
2023-03-10T16:58:50.6423139Z     result.compute()
2023-03-10T16:58:50.6423532Z .nox/tests/lib/python3.9/site-packages/dask/base.py:314: in compute
2023-03-10T16:58:50.6423899Z     (result,) = compute(self, traverse=False, **kwargs)
2023-03-10T16:58:50.6424333Z .nox/tests/lib/python3.9/site-packages/dask/base.py:599: in compute
2023-03-10T16:58:50.6424664Z     results = schedule(dsk, keys, **kwargs)
2023-03-10T16:58:50.6425061Z .nox/tests/lib/python3.9/site-packages/dask/threaded.py:89: in get
2023-03-10T16:58:50.6425362Z     results = get_async(
2023-03-10T16:58:50.6425770Z .nox/tests/lib/python3.9/site-packages/dask/local.py:511: in get_async
2023-03-10T16:58:50.6426091Z     raise_exception(exc, tb)
2023-03-10T16:58:50.6426494Z .nox/tests/lib/python3.9/site-packages/dask/local.py:319: in reraise
2023-03-10T16:58:50.6426777Z     raise exc
2023-03-10T16:58:50.6427148Z .nox/tests/lib/python3.9/site-packages/dask/local.py:224: in execute_task
2023-03-10T16:58:50.6427473Z     result = _execute_task(task, data)
2023-03-10T16:58:50.6427876Z .nox/tests/lib/python3.9/site-packages/dask/core.py:119: in _execute_task
2023-03-10T16:58:50.6428508Z     return func(*(_execute_task(a, cache) for a in args))
2023-03-10T16:58:50.6428964Z .nox/tests/lib/python3.9/site-packages/dask/array/core.py:4394: in store_chunk
2023-03-10T16:58:50.6429357Z     return load_store_chunk(x, out, index, lock, return_stored, False)
2023-03-10T16:58:50.6429829Z .nox/tests/lib/python3.9/site-packages/dask/array/core.py:4376: in load_store_chunk
2023-03-10T16:58:50.6430144Z     out[index] = x
2023-03-10T16:58:50.6430456Z lib/iris/fileformats/netcdf/saver.py:521: in __setitem__
2023-03-10T16:58:50.6430824Z     dataset = _thread_safe_nc.DatasetWrapper(self.path, "r+")
2023-03-10T16:58:50.6431199Z lib/iris/fileformats/netcdf/_thread_safe_nc.py:55: in __init__
2023-03-10T16:58:50.6431532Z     instance = self.CONTAINED_CLASS(*args, **kwargs)
2023-03-10T16:58:50.6431896Z src/netCDF4/_netCDF4.pyx:2466: in netCDF4._netCDF4.Dataset.__init__
2023-03-10T16:58:50.6432210Z     ???
2023-03-10T16:58:50.6432485Z src/netCDF4/_netCDF4.pyx:1615: in netCDF4._netCDF4._get_format
2023-03-10T16:58:50.6432761Z     ???
2023-03-10T16:58:50.6433000Z _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
2023-03-10T16:58:50.6433173Z 
2023-03-10T16:58:50.6433245Z >   ???
2023-03-10T16:58:50.6433497Z E   RuntimeError: NetCDF: Not a valid ID
2023-03-10T16:58:50.6433667Z 
2023-03-10T16:58:50.6433804Z src/netCDF4/_netCDF4.pyx:2028: RuntimeError
```

pp-mo commented:
@bouweandela do you have time to run your checks through this, as you did with the original #5031 ?

I know the tests still aren't all passing, but it's only code coverage warnings.
We will ceratainly improve that, by adding testing, before we're done with this.

fnattino commented:
Hi @pp-mo, thank you so much for all your work on this and sorry for taking long to getting back to you! I have been running a couple of tests using either [ESMValCore@dask-distributed](https://github.com/ESMValGroup/ESMValCore/tree/dask-distributed) (only with the distributed scheduler) or using some examples derived from your gist (both multithreading and distributed) - it seems to be working very nicely! 

It seems indeed that the distributed scheduler cannot be used with the direct streaming of variable data, because of the file already open. I had a look at how Xarray does `to_netcdf` with `compute=True`, and it looks to be doing things exactly as in your current implementation: keeping reference to sources and targets, then closing the file (if using distributed) and only then running `da.store`. 

For what concerns the fill-value checks, one could maybe set them up as additional dask tasks on the source? One could pass the `fill_value` to `_deferred_save` and then do something like:
```python
            sources, targets = zip(*self.deferred_writes)
            result = da.store(sources, targets, compute=False, lock=False)

            @dask.delayed
            def check_fill_value(arr):
                return fill_value in arr if fill_value is not None else False

            @dask.delayed
            def raise_warning(filename, *contains_fill_value):
                if any(contains_fill_value):
                    # to warn with distributed maybe do like: https://github.com/dask/dask/issues/9393#issue-1342083979 
                    warnings.warn(f"fill-value collision in {filename}")

            warns = []
            for source, target in self.deferred_writes:
                contains_fill_value = [check_fill_value(chunk) for chunk in source.to_delayed().flatten()]
                warns.append(raise_warning(target.varname, *contains_fill_value))
```
It is true that the checks are not really carried out while storing, but maybe this is a way to keep these (and the warnings)? As it is right now, the code snippet requires then to call `dask.compute()` (or `client.compute()`) with both `result` and `warns` as arguments, in order to trigger storing & warnings, but this could be probably improved.

pp-mo commented:
@fnattino thanks for your input !
 
 > I had a look at how Xarray does `to_netcdf` with `compute=True` 

Yes.  Like a lot of this code, I started out saying I didn't want to replicate how Xarray does it, but ended up doing exactly that! 
I suspect one day we may want to have open file caching like they do it, too.

> For what concerns the fill-value checks, one could maybe set them up as additional dask tasks on the source? 

Thanks that's a really neat idea!!!
I may well pursue this ...   As I'm already constructing a special deferred-store object, there's no reason I can't add more into that operation.  

> It is true that the checks are not really carried out while storing,
However IIUC the magic of Dask _ought_ to at least re-use each chunk for both operations without re-loading.  Just checking a visualisation graph should confirm that -- I'll try it.

My real concern with this fill-value problem was not that the existing behaviour is totally crucial for us, but we want to solve 2 problems at once : 
  1. preserve existing behaviour with fill-value checking (which the current code here _is not doing_), and 
  2. avoid making the save code dependent on which scheduler

Doing _both_ those together is the hard bit where I got stuck!  I think your idea probably lets us fix it, so I'll definitely try  that.  
Thanks again! :brain: :1st_place_medal:

fnattino commented:
> It is true that the checks are not really carried out while storing,
However IIUC the magic of Dask ought to at least re-use each chunk for both operations without re-loading. Just checking a visualisation graph should confirm that -- I'll try it.

I did check the graph for a test derived from your example, and it seems indeed that the chunks are not reloaded: 

![mydask](https://user-images.githubusercontent.com/49899980/225587206-346d3f3f-cbe0-4aff-a0a4-b67a1cf84530.png)

Actually, I think that reloading cannot really happen in the current form of the code, since the targets that are passed to `da.store` are actually  `DeferredSaveWrapper` objects, which have `__setitem__` but not `__getitem__`..

bouweandela commented on lib/iris/fileformats/netcdf/saver.py:
```diff
@@ -2652,6 +2715,14 @@ def save(
+    * compute (bool):
+        When False, create the output file but defer writing any lazy array content to
+        its variables, such as (lazy) data and aux-coords points and bounds.
+        Instead return a class:`dask.delayed` which, when computed, will compute all
```
```suggestion
        Instead return a class:`dask.delayed.Delayed` which, when computed, will compute all
```

bouweandela reviewed (commented):
Great to see progress on this! I just read through the code and noticed some small typos so thought I'd help with correcting those — no real comments from my side.

pp-mo commented on lib/iris/fileformats/netcdf/saver.py:
```diff
@@ -2455,16 +2474,37 @@ def _lazy_stream_data(data, fill_value, fill_warn, cf_var):
+                def store(data, cf_var, fill_value):
+                    # Create a data-writeable object that we can stream into, which
+                    # encapsulates the file to be opened + variable to be written.
+                    writeable_var_wrapper = (
+                        _thread_safe_nc.DeferredSaveWrapper(
+                            self.filepath, cf_var, self.file_write_lock
+                        )
+                    )
```
> reloading cannot really happen in the current form of the code, since the targets that are passed to da.store are actually DeferredSaveWrapper objects

I was thinking in terms of the **_input_** chunks, not the output ones -- the relationship may not always be 1-1 of course.

It's down to dask to only compute each input chunk once.  However, I believe that this is a safe bet, as Dask probably treats the array accesses as generic "tasks", and does not consider its tasks to be re-rerunnable.
So, it is always doing some results caching + re-use.   What is **_not_** so clear, in my experience, is whether it always _discards_ cached results when it could + should.  This is obviously critical to the memory performance of store operations, and I has definitely caused us some problems before.

pp-mo replied:
This is a product of black.
So 0 arguments but 0 control over the result !!

I can get rid of the extra parens by making it less wide -- I'll fix the import + naming strategy.

pp-mo commented:
> @fnattino reloading cannot really happen in the current form of the code, since the targets that are passed to da.store are actually DeferredSaveWrapper objects

I was thinking in terms of the **_input_** chunks, not the output ones -- the relationship may not always be 1-1 of course.

It's down to dask to only compute each input chunk once.  However, I believe that this is a safe bet, as Dask probably treats the array accesses as generic "tasks", and does not consider its tasks to be re-rerunnable.
So, it is always doing some results caching + re-use.   What is **_not_** so clear, in my experience, is whether it always _discards_ cached results when it could + should.  This is obviously critical to the memory performance of store operations, and has definitely caused us some problems before.

fnattino commented on lib/iris/fileformats/netcdf/_thread_safe_nc.py:
```diff
@@ -340,3 +340,41 @@ def __getstate__(self):
             setattr(self, key, value)
+
+
+class DeferredSaveWrapper:
```
This has been created as a way to set up the deferred writing, but right now this class is not really doing any "deferred" action, right? So maybe just `SaveWrapper` or something like `NetCDFDataWriter`?

pp-mo commented:
Modified to perform delayed fill-value checks.
I also resolved conflicts in the lockfiles here in GitHub, just to see how the tests would run.  That  might have been a mistake ..

pp-mo commented:
Re-merged + fixed lockfiles *again*

pp-mo commented:
Note : merged back from main to resolve conflicts.
Remade lockfiles, since this still has different requirements --  pinning libnetcdf<4.9 (not yet in main).
Also fixed Makefile,  which also needs doing on main if we are keeping it ??

pp-mo commented:
I think this is now ready for review at last.
I have now added all the tests I wanted, including testing operation with the dask distributed scheduler.

I am still considering where best to document the feature, though.
Not quite sure where it goes, apart from the whatsnew.  Possibly #5190 may help ?

pp-mo commented:
Note on "what remains"
A few outstanding bullet points from the header comment, also some possible additions  .
  - **whatsnew entry** : still to-do
  - ?further documentation? :  still considering where to put this + how much -- also relates to whats-new
  - **benchmark to demonstrate performance** : possible
  - **allowing a lock to be passed in by the user** : 
  - ?more unit-testing? : it might be appropriate to add some unit-testing of `Saver.delayed_completion` and `Saver.complete`.  
    This might improve the coverage score (e.g. complete has untested error code).  But the way the function behaviours inter-operate makes isolated testing hard.

pp-mo commented:
> Allowing a lock to be passed in by the user

As suggested by @fnattino on the original PR.

I find I now have a bit of a problem with this suggestion + want to **_propose we don't do it_**.

For background ...
Our Iris  implementation effectively uses one common lock (netcdf file lock) for all file accesses, *plus* a file-specific lock for writes.  
Whereas, Xarray uses its ["combined" locks](https://github.com/pydata/xarray/blob/v2023.03.0/xarray/backends/locks.py#L195-L210) for this :  So, a dataset is given a single lock when opened for read or write :  if read, this is just the the default lock the common [netcdf-library-lock](https://github.com/pydata/xarray/blob/v2023.03.0/xarray/backends/netCDF4_.py#L46), and if write it is a [combination of this with a file-specific write lock](https://github.com/pydata/xarray/blob/v2023.03.0/xarray/backends/netCDF4_.py#L374). 
(And note that the 'netcdf-library' lock is already a combination with a separate lock for the HDF5 module, since xarray can use that independently of netCDF).

However, there seems to be a big hole in the current Xarray strategy since, although the underlying [`NetCDF4DataStore` supports passing in a lock on creation](https://github.com/pydata/xarray/blob/v2023.03.0/xarray/backends/netCDF4_.py#L46), you cannot pass a lock into the `to_netcdf` operation -- nor can you pass in a datastore, but it always [creates one itself, without passing any user-specific controls](https://github.com/pydata/xarray/blob/v2023.03.0/xarray/backends/api.py#L1172-L1200).  
So, **I don't think this strategy is complete, and we're not ready to align Iris with it.**

I think we probably _will_ need to address these issues in #4994 (to make iris and xarray fetches co-operate), but I'm happy that for such obscure uses we can for now manage this by patching, and _we aren't ready to propose a fully-realised public solution_, especially if Xarray is also missing functionality in this area.

fnattino commented:
Hi @pp-mo, thanks a lot for all the work on this! I should have time to have a look and run few tests by the end on the week, I will keep you posted.

pp-mo commented:
> [Documentation section and whatsnew entry.](https://github.com/SciTools/iris/pull/5191/commits/7daee68a464f0b1689f79d1d26187a7de59065f7)

Actually I'm not quite done with this.
I think I need to explain that we are also now supporting operations with 'distributed'.
It wasn't clear before, but we couldn't save to netcdf with a distributed dask scheduler.
Which now we can (and it's [tested](https://github.com/SciTools/iris/blob/97474f9bbcc3e29f2e4906b2599e802f7db82e96/lib/iris/tests/integration/netcdf/test_delayed_save.py#L209-L258) !).

pp-mo commented:
Ok I hope this is now adequate for the documentation.
I have added a section to the User Guide, but it seems over-detailed to me for that context.
I think we should consider moving it in future to a 'further topics' sections with the [forthcoming "dask best practices"](#5190).

lbdreyer commented:
> > The new user guide docs aren't displaying in the docs build: https://scitools-iris--5191.org.readthedocs.build/en/5191/userguide/real_and_lazy_data.html#delayed-netcdf-saving although i
> 
> ?? but this same link is **_working_** for me ?? Possibly a refresh issue. Can you check it out @lbdreyer ?

Tried in a different browser and it is now working! So it looks like it was a refresh issue

lbdreyer commented on docs/src/whatsnew/latest.rst:
```diff
@@ -53,6 +53,10 @@ This document explains the changes made to Iris for this release
    :pull:`4972`)
 
+#. `@pp-mo`_ supported delayed saving of lazy data, when writing to the netCDF file
+   format.  See : :ref:`delayed netCDF saves <delayed_netcdf_save>`.
```
```suggestion
   format.  See :ref:`delayed netCDF saves <delayed_netcdf_save>`.
```
Has something gone wrong with your formatting here?

pp-mo replied:
Not exactly, I meant "See : <link>", but the extra colon is not really  needed -- will remove.

lbdreyer commented on docs/src/userguide/real_and_lazy_data.rst:
```diff
@@ -228,10 +230,47 @@ coordinates' lazy points and bounds:
+thus enabling parallel save operations.
+
+This works in the following way :
+    1. an :func:`iris.save` call is made, with a netcdf file output and the additional
```
```suggestion
    1. an :func:`iris.save` call is made, with a netCDF file output and the additional
```
I believe NetCDF/netCDF is the official name. As this is part of the docs (so user facing) we should probably use the official name/correct capitalisation

lbdreyer reviewed (changes requested):
I've not completed looking through this PR as there are some parts I want to take another look at. But in the meantime, I wanted to submit my comments so far, in case you wanted to make a start on addressing them.

Overall this is looking really good! Most of my comments of very minor as you will see below

pp-mo commented:
> @fnattino I should have time to have a look and run few tests by the end on the week, I will keep you posted.

Hi @fnattino @bouweandela , any progress on confirming whether the latest version is working for you ?

pp-mo commented on lib/iris/tests/unit/fileformats/netcdf/saver/saver/__init__.py:
```diff
@@ -0,0 +1,6 @@
+# This file is part of Iris and is released under the LGPL license.
+# See COPYING and COPYING.LESSER in the root of the repository for full
+# licensing details.
+"""Unit tests for the :class:`iris.fileformats.netcdf.saver.Saver` class."""
```
:+1:
I see that we have not named test directories after classes before, at all.  So it's not clear if this is a good idea.
On reflection, it's also not really needed.  Instead I think I should keep the directories organised by package only, and  just rename some things 
  * `tests/unit/fileformats/netcdf/saver/saver/test_Saver<*>.py` --> 
     `tests/unit/fileformats/netcdf/saver/test_Saver<*>.py`
  * `tests/unit/fileformats/netcdf/saver/saver/test__lazy_stream_data.py` -->
     `tests/unit/fileformats/netcdf/saver/saver/test_Saver__lazy_stream_data.py`

pp-mo commented on lib/iris/tests/unit/fileformats/netcdf/saver/test_save.py:
```diff
@@ -3,31 +3,32 @@
 
+# Import iris.tests first so that some things can be initialised before
+# importing anything else.
+import iris.tests as tests  # isort:skip
```
Mistake, somehow !

pp-mo commented on lib/iris/tests/integration/netcdf/test_delayed_save.py:
```diff
@@ -0,0 +1,258 @@
+        if not save_is_delayed:
+            result_warnings = [log.message for log in logged_warnings]
+        else:
+            assert len(logged_warnings) == 0
```
Ok, good point.
I think the nicest way is to make these warnings distinguishable by giving them a specific subclass.
I'll do that !

pp-mo commented on lib/iris/tests/integration/netcdf/test_delayed_save.py:
```diff
@@ -0,0 +1,258 @@
+    def getmask(cube_or_coord):
+        cube_or_coord = cube_or_coord.copy()  # avoid realising the original
+        if hasattr(cube_or_coord, "points"):
+            data = cube_or_coord.points
```
Ok I have added those too.

pp-mo commented:
Thanks @lbdreyer
I hope the recent changes address the points you raised so far.
Please re-review !

fnattino commented on lib/iris/tests/integration/netcdf/test_delayed_save.py:
```diff
@@ -0,0 +1,258 @@
+
+    @classmethod
+    @pytest.fixture(params=["ThreadedScheduler", "DistributedScheduler"])
+    def scheduler_type(cls, request):
+        sched_typename = request.param
+        if sched_typename == "ThreadedScheduler":
+            config_name = "threads"
+            if cls._distributed_client is not None:
+                cls._distributed_client.close()
+                cls._distributed_client = None
+        else:
+            config_name = "distributed"
+            if cls._distributed_client is None:
+                cls._distributed_client = distributed.Client()
+
+        with dask.config.set(scheduler=config_name):
+            yield sched_typename
```
Running tests locally, I had issues with this one, I think because the distributed scheduler would not be properly cleaned up at the end of the last test. The following changes make sure the distributed scheduler is closed at the end of each test, and it solves the failing tests I got:
```suggestion
    def scheduler_type(cls, request):
        sched_typename = request.param
        if sched_typename == "ThreadedScheduler":
            config_name = "threads"
        else:
            config_name = "distributed"
            if cls._distributed_client is None:
                cls._distributed_client = distributed.Client()

        with dask.config.set(scheduler=config_name):
            yield sched_typename
        if cls._distributed_client is not None:
            cls._distributed_client.close()
            cls._distributed_client = None
```

pp-mo replied:
@fnattino thanks,  I see it now !
I really did get this wrong -- we don't want to leave a distributed client in existence after any of these tests, we should always destroy it.
Frankly I'm really surprised this didn't upset other tests within the CI. 
I will fix ...

lbdreyer reviewed (changes requested):
A few more minor comments I noticed

pp-mo commented on lib/iris/fileformats/netcdf/saver.py:
```diff
@@ -2454,64 +2341,177 @@ def _lazy_stream_data(data, fill_value, fill_warn, cf_var):
+            # save lazy data with a delayed operation.  For now, we just record the
+            # necessary information -- a single, complete delayed action is constructed
+            # later by a call to delayed_completion().
+            def store(data, cf_var, fill_value):
+                # Create a data-writeable object that we can stream into, which
+                # encapsulates the file to be opened + variable to be written.
+                write_wrapper = _thread_safe_nc.NetCDFWriteProxy(
+                    self.filepath, cf_var, self.file_write_lock
+                )
+                # Add to the list of delayed writes, used in delayed_completion().
+                self._delayed_writes.append((data, write_wrapper, fill_info))
+                # In this case, fill-value checking is done later. But return 2 dummy
+                # values, to be consistent with the non-streamed "store" signature.
+                is_masked, contains_value = False, False
+                return is_masked, contains_value
+
+        else:
+            # Real data is always written directly, i.e. not via lazy save.
+            # We also check it immediately for any fill-value problems.
+            def store(data, cf_var, fill_value):
+                cf_var[:] = data
+                return _data_fillvalue_check(np, data, fill_value)
```
OK, I think I originally wrote it this slightly peculiar way for continuity with the existing code.
So there were **_three_** possible 'store' routines in the original for of this code, and they were at a lower-level than the 'delayed_save' code, taking just a fill_value arg, and not the whole fill-info (which is a new idea). 

But now I agree, what you suggest is simpler to understand -- especially as the deferred form doesn't actually _use_ the passed fill-value !

pp-mo commented on requirements/locks/py310-linux-64.lock:
OK I'll do this again when I'm happy with the current review  changes.

pp-mo commented on lib/iris/fileformats/netcdf/_dask_locks.py:
```diff
@@ -0,0 +1,133 @@
+types, i.e. local 'threads' scheduler, local 'processes' or distributed.
+
+In any case, an "iris.fileformats.netcdf.saver.Saver" object contains a netCDF4.Dataset
+targetting an output file, and creates a Saver.file_write_lock object to serialise
```
I reckon this is a good spelling + actually prefer it !
-- noted that GitHub here does consider either valid.

pp-mo commented:
Thanks @fnattino @lbdreyer .
Hope I've addressed most of the above.
I'm pushing at this point to re-run tests, hoping that fixing the use of a distributed client as suggested by @fnattino **_might_** fix the outstanding test failure, since I didn't have that failure locally.  If not, I'm not really sure what could be going wrong there + it will need investigating...

To-dos outstanding from @lbdreyer suggestions:
  * re-resolve lockfiles
  * add unit testing for `iris.fileformats.netcdf._dask_locks`

pp-mo commented:
> hoping that fixing the use of a distributed client as suggested by @fnattino might fix the outstanding test failure, since I didn't have that failure locally. If not, I'm not really sure what could be going wrong there + it will need investigating...

**Please bear with me while I look into this**

Something very odd + worrying is happening here.
This failure seems consistent, but I can't reproduce it locally even in an env built from the lockfile, or running via nox.

pp-mo commented on lib/iris/fileformats/netcdf/saver.py:
```diff
@@ -2454,64 +2345,175 @@ def _lazy_stream_data(data, fill_value, fill_warn, cf_var):
+            Any warnings that were raised while writing delayed data.
+
+        """
+        if self._dataset._isopen:
```
```suggestion
        if self._dataset.isopen():
```
Just spotted : this is not good usage, use the public API form instead.

pp-mo commented:
Sorry I had to rebase, because the history had become unmanageable and I wanted to squash some commits, but it required manual resolution.  
I have had 4-5 times today to re-make the lockfiles and resolve conflicts in the pyXX.yml and latest.rst !

Still to-do:
  - [ ] replace [dodgy use of `Dataset_isopen`](https://github.com/SciTools/iris/pull/5191/#discussion_r1165282708)
  - [ ] add specific testing for content of `iris.fileformats.netcdf._dask_locks`
     * NB - this has also acquired knowledge to recognise "single-threaded" scheduler.
  - [ ] review testing against different scheduler types -- possibly now overkill ?
  - [ ] review other places where we might have [crossover between tests because of using a common filepath](https://github.com/SciTools/iris/pull/5191/commits/0da68cfe635605bf1cae808e30d61720a34d008c)

pp-mo commented:
Ok, now removed from draft.

I found that the basic problem with the 'time_of_writing' tests was that the small cell-measures/ancils added had been loaded as real data due to #5069 : Fixed by patching the loader and added checks that loaded components are lazy.
I still haven't fully understood how these were passing locally, but I think it may have been due (?somehow) due to re-using the same output testfile path for all the tests, so there was some kind of 'cross-talk'.

pp-mo commented:
I will proceed to "add [specific testing for content of iris.fileformats.netcdf._dask_locks](https://github.com/SciTools/iris/pull/5191/files#r1163363655)"

@lbdreyer  I hope this is OK.
Please do consider whether you think the testing of the new Saver routines is adequate -- there is no unit testing of `delayed_completion` or `complete` methods (unlike `lazy_stream_data`), mostly because I reckoned it was simply too hard to do given the complex exection context and the coupling between them (!)
Behaviour should be tested well enough via the integration tests. I think.

pp-mo commented:
Ok, possibly that is now ok to go.
@lbdreyer please re-review !

bouweandela commented:
@pp-mo @trexfeathers @ESadek-MO I noticed the first release candidate of v3.5 has been cut. Do you reckon there is still a chance that this may be included in v3.5?

bjlittle commented:
@bouweandela Note that we're pulling forward the release of `iris` 3.6 to `June` instead of `September`, so delayed saving may land in time for this release.

I'm sure that @pp-mo can give a fuller update on our expectation of when this will be available 👍

pp-mo commented:
> @pp-mo @trexfeathers @ESadek-MO I noticed the first release candidate of v3.5 has been cut. Do you reckon there is still a chance that this may be included in v3.5?

Sincere apologies, we had a few remaining problems with this code in our recent work and it is still not merged.
So it just missed the deadline for Iris3.5.

I think the problems are now resolved, and we expect to get it completed + merged quite soon now. 
So then it will be in main, but obviously to appear in a release, it now has to wait for Iris3.6.

I've put this at No.1 on the Iris 3.6 board.
Plans for 1.6 : 
  * there are already a few relatively minor changes tasked there
     ( mostly to enable #4994 , but N.B. the bulk of that will now be delivered _outside_ Iris. )
  * we intend to get v1.6 out ASAP once those things are dealt with, 
    * _not_ to wait for an arbitrary timepoint, and 
    * _not_ to add anything major to it.  
  * Bigger things can be pushed to v1.7 (which is itself [scheduled for August] (https://github.com/SciTools/iris/discussions/5209#discussion-5000919)).

Apologies again for the delay @bouweandela.  We _could_ make a special release for this, but I'm hoping that won't be necessary for your requirements ?

pp-mo commented:
> Sincere apologies, we had a few remaining problems with this code in our recent work and it is still not merged. So it just missed the deadline for Iris3.5.

Apologies for spurious mention of "1.6" and "1.7" releases -- now fixed !

bouweandela commented:
> @bouweandela Note that we're pulling forward the release of iris 3.6 to June instead of September, so delayed saving may land in time for that release. See https://github.com/SciTools/iris/discussions/5106

Great to see that the release pace of iris is increasing @bjlittle! The reduced waiting time for new features will definitely make it more attractive to contribute and easier to use.

> Apologies again for the delay @bouweandela. We could make a special release for this, but I'm hoping that won't be necessary for your requirements ?

Unfortunately the [ESMValTool release schedule](https://docs.esmvaltool.org/en/latest/community/release_strategy/release_strategy.html#release-schedule) currently has a release in June and the next release in October, so then our October release would be the first release where our users could benefit from the new feature, which would be disappointingly late. Let me check with the @ESMValGroup/esmvaltool-coreteam though, maybe we could postpone our release by a month. How sure are you that your next release will not be pushed back further? We would need at least 2 weeks between the iris release with this feature and our release for testing.

bouweandela commented:
> **Note on PR checking status:**
There is really no point my chasing my tail to re-update latest.rst and lockfiles every time something new hits main.
I have already done this 3-4 times.

I think this could be remedied by setting up a [merge queue](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/managing-a-merge-queue), maybe that could be an interesting feature for you?

bjlittle commented:
> Great to see that the release pace of iris is increasing @bjlittle! The reduced waiting time for new features will definitely make it more attractive to contribute and easier to use.

This is great to hear. Thanks for the feedback, it makes a big difference :+1: 

> Unfortunately the [ESMValTool release schedule](https://docs.esmvaltool.org/en/latest/community/release_strategy/release_strategy.html#release-schedule) currently has a release in June and the next release in October, so then our October release would be the first release where our users could benefit from the new feature, which would be disappointingly late. Let me check with the [@ESMValGroup/esmvaltool-coreteam](https://github.com/orgs/ESMValGroup/teams/esmvaltool-coreteam) though, maybe we could postpone our release by a month. How sure are you that your next release will not be pushed back further? We would need at least 2 weeks between the iris release with this feature and our release for testing.

Again, thanks for sharing. This is really useful to know. It gives me the context I need to make some decisions about the next release.

Just so you know, we have a **2 week** (9 days only, thanks to the [bank holiday for King Charles III coronation](https://www.gov.uk/government/news/bank-holiday-proclaimed-in-honour-of-the-coronation-of-his-majesty-king-charles-iii)) development sprint scheduled, dedicated to the `iris` 3.6 release from `Tue 09 May` to `Fri 19 May`. I'm very keen to keep this release **extremely** light in content, that way we can guarantee (as best as possible) to be nimble and not slip delivery to the right. I'm going to be pretty strict about that.

I just need to get a clear understanding from our side about the absolute minimal features that we want to include for the `iris` 3.6 release. At that point I'd really like to hangout with yourself @bouweandela (and anyone else from `esmvaltool`) to get perspective on what content you need and by when. Does that sound like a reasonable plan?

I see that your forthcoming `esmvalcore` 2.9.0 feature freeze is on `Mon 05 Jun`. That would mean you need an `iris` release candidate available on `Mon 22 May` for your 2 weeks of testing.

It would be interesting to see if we could make that happen for you guys. It's quite tight, but not impossible, and pretty much all depends on content and resourcing.

Leave it with me. I'll definitely be in touch very soon :wink:

trexfeathers commented:
>I think this could be remedied by setting up a [merge queue](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/managing-a-merge-queue), maybe that could be an interesting feature for you?

I believe that would still require @pp-mo to be aware of which pull requests affect `latest.rst` of `requirements/locks/` so that the queue can be set up. And presumably if this happens multiple times, as mentioned, it would require a new queue to be set up each time.

pp-mo commented:
> > Great to see that the release pace of iris is increasing @bjlittle!

> Leave it with me. I'll definitely be in touch very soon 😉

Thanks people.
It does feel great to be discussing actual priority + scheduling with users !

lbdreyer reviewed (approved):
I have no outstanding comments so I believe this is now ready to merge!

@pp-mo could you sort out the lockfiles/whatsnew then ping me to let me now that I can merge it in?

pp-mo commented:
Phew 😅 
A lot of fuss today fixing conflicts + knockons due to activity on the main branch 
-- if only I'd done it yesterday !

@lbdreyer please re-review when you can

lbdreyer commented:
Merged :partying_face: with quite possibly the l[ongest commit message](https://github.com/SciTools/iris/commit/94e44efead1fe2a52e5267e94c6286c74449680a) I've witnessed

bouweandela commented:
Thanks a lot for all the work on this!
